### PR TITLE
Moved to GH actions; Added lint; Added issue/PR templates.

### PR DESCRIPTION
**Pull request by bwplotka**
Also pinned Go tools thanks to https://github.com/bwplotka/bingo

Signed-off-by: Bartlomiej Plotka <bwplotka@gmail.com>

### CODE REVIEW

**Comment by bwplotka** [comment]
Let's make sure the CI GH actions runs here

File: .github/stale.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+pull:
+  daysUntilClose: 14
+  daysUntilStale: 30
+  markComment: >
+    Hello 👋 Looks like there was no activity on this amazing PR for last 1 month.
+
+    **Do you mind updating us on the status?** Is there anything we can help with? If you plan to still work on it, just comment on this PR or push a commit. Thanks! 🤗
+
+    If there will be no activity for next 2 weeks, this issue will be closed (we can always reopen a PR if you get back to this!).
+    Alternatively, use [`remind` command](https://probot.github.io/apps/reminders/) if you wish to be reminded at some point in future.
```
Comment (johanbrandhorst) [inline .github/stale.yml:32]:
```suggestion
    Alternatively, use [the `remind` command](https://probot.github.io/apps/reminders/) if you wish to be reminded at some point in the future.
```

File: .github/stale.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+issues:
+  daysUntilClose: 14
+  daysUntilStale: 60
+  markComment: >
+    Hello 👋 Looks like there was no activity on this issue for last 2 months.
+
+    **Do you mind updating us on the status?** Is this still reproducible or needed? If yes, just comment on this PR or push a commit. Thanks! 🤗
+
+    If there will be no activity for next 2 weeks, this issue will be closed (we can always reopen an issue if we need!).
+    Alternatively, use [`remind` command](https://probot.github.io/apps/reminders/) if you wish to be reminded at some point in future.
```
Comment (johanbrandhorst) [inline .github/stale.yml:46]:
```suggestion
    Alternatively, use [the `remind` command](https://probot.github.io/apps/reminders/) if you wish to be reminded at some point in the future.
```

File: Makefile [hunk 3]
```diff
@@ -1,37 +1,75 @@
...
+.PHONY: fmt
+fmt:
+	@echo "Running fmt for all modules: ${MODULES}"
+	@$(GOIMPORTS) -local github.com/grpc-ecosystem/go-grpc-middleware/v2 -w $(GOFILES_NOVENDOR)
+	@$(foreach dir,$(MODULES),$(MAKE) fmt_module DIR=$(dir))
 
-all: vet fmt test
+.PHONY: fmt_module
+fmt_module:
+	go fmt $(shell cd $(DIR) && go list ./... | grep -v /vendor/)
```
Comment (johanbrandhorst) [inline Makefile:24]: We don't have a vendor directory anymore, maybe we could simplify this?
```suggestion
	gofmt -w -s ./...
```

File: Makefile [hunk 4]
```diff
@@ -1,37 +1,75 @@
...
-proto: ./grpctesting/testpb/test.proto
-	@scripts/genproto.sh
-
-# TODO(bwplotka): This depends on test_proto, but CI does not have it, so let's skip it for now.
-test: vet
-	cd ./providers/kit/ && go test -v -race ./...
-	cd ./providers/logrus/ && go test -v -race ./...
-	cd ./providers/zap/ && go test -v -race ./...
+.PHONY: proto
+proto: $(PROTOC_GEN_GOGOFAST) $(GOIMPORTS) $(PROTOC) ./grpctesting/testpb/test.proto
```
Comment (johanbrandhorst) [inline Makefile:27]: Unrelated to this PR, but we should be able to switch to `google.golang.org/protobuf` now.

File: Makefile [hunk 5]
```diff
@@ -1,37 +1,75 @@
...
+	@GOIMPORTS_BIN="$(GOIMPORTS)" PROTOC_BIN="$(PROTOC)" scripts/genproto.sh
+
+.PHONY: test
+test:
+	@echo "Running tests for all modules: ${MODULES}"
+	@$(foreach dir,$(PROVIDER_MODULES),$(MAKE) test_module DIR=$(dir))
```
Comment (johanbrandhorst) [inline Makefile:33]: What does this even do?

File: Makefile [hunk 6]
```diff
@@ -1,37 +1,75 @@
...
+# Add
+#      --cpu-profile-path string   Path to CPU profile output file
+#      --mem-profile-path string   Path to memory profile output file
+# to debug big allocations during linting.
+lint: ## Runs various static analysis against our code.
+lint_module: $(FAILLINT) $(GOLANGCI_LINT) $(MISSPELL) fmt_module
+	$(call require_clean_work_tree,"detected not clean master before running lint")
+	@echo ">> verifying modules being imported"
+	@$(FAILLINT) -paths "errors=github.com/pkg/errors" ./...
+	@$(FAILLINT) -paths "fmt.{Print,PrintfPrintln,Sprint}" -ignore-tests ./...
```
Comment (johanbrandhorst) [inline Makefile:56]: Can we merge these runs

File: Makefile [hunk 7]
```diff
@@ -1,37 +1,75 @@
...
+	@echo ">> examining all of the Go files"
+	@go vet -stdmethods=false ./...
+	@echo ">> linting all of the Go files GOGC=${GOGC}"
+	@$(GOLANGCI_LINT) run
+	@echo ">> detecting misspells"
+	@find . -type f | grep -v vendor/ | grep -vE '\./\..*' | xargs $(MISSPELL) -error
```
Comment (johanbrandhorst) [inline Makefile:62]: No vendor directory

File: Makefile [hunk 8]
```diff
@@ -1,37 +1,75 @@
...
+	@echo ">> detecting white noise"
+	@find . -type f \( -name "*.md" -o -name "*.go" \) | SED_BIN="$(SED)" xargs scripts/cleanup-white-noise.sh
```
Comment (johanbrandhorst) [inline Makefile:64]: Do we need this?

File: Makefile (hunk 5 above)
Comment (bwplotka) [reply Makefile:33]: This ensures we support multiple nested modules fun! (:

File: Makefile (hunk 6 above)
Comment (bwplotka) [reply Makefile:56]: Do you want to block fmts in tests? then yes (:

File: Makefile (hunk 7 above)
Comment (bwplotka) [reply Makefile:62]: Well, but someone can download vendor locally and it should work IMO (:

File: Makefile (hunk 8 above)
Comment (bwplotka) [reply Makefile:64]: yea I constantly commit white noise everywhere (: Up to you thoguh

**Comment by bwplotka** [comment]
TODO still GH err: https://github.com/grpc-ecosystem/go-grpc-middleware/actions/runs/119943153

File: Makefile (hunk 8 above)
Comment (johanbrandhorst) [reply Makefile:64]: I mean, I don't know what it's doing. Doesn't `gofmt` format things enough?

File: Makefile (hunk 7 above)
Comment (johanbrandhorst) [reply Makefile:62]: Really? Our Makefile needs to cater to users cloning this _and_ running `go mod vendor` _and_ CI still working?

File: Makefile (hunk 6 above)
Comment (johanbrandhorst) [reply Makefile:56]: I suppose we can use `t.Logf` in tests? Also, should it be
```suggestion
	@$(FAILLINT) -paths "errors=github.com/pkg/errors" ./...
	@$(FAILLINT) -paths "fmt.{Print,Printf,Println,Sprint}" -ignore-tests ./...
```

Comment (bwplotka) [reply Makefile:56]: Can agree (:

File: Makefile (hunk 7 above)
Comment (bwplotka) [reply Makefile:62]: Yea, you can blame me I do that a lot =D

No strong opinion we can start with your idea first, will remove :hug

File: Makefile (hunk 8 above)
Comment (bwplotka) [reply Makefile:64]: Nope. 

It's mainly for markdown files
It's oven the case ppl leave "Dead" whitespace (after a line, or empty line have some whitespace, of rile does not end with newline etc)

**Comment by bwplotka** [comment]
Super thank your for awesome quick review, will address this afternoon! (:

File: .github/stale.yml (hunk 1 above)
Comment (bwplotka) [reply .github/stale.yml:32]: actually no remind is set up so let's close it for now

**Comment by irridia** [comment]
Be sure to update the import paths in the examples_test files, both for users and it confuses `go mod`.

**Comment by bwplotka** [comment]
Good point, thanks

>

**Comment by bwplotka** [comment]
Lot's of vert / lint issue to go through...

**Comment by bwplotka** [comment]
Should be fine now, PTAL @yashrsharma44 @johanbrandhorst

**Comment by bwplotka** [comment]
Close to be finished..  two last standing thigs:

1. `make lint` works locally but not on GH Actions:

```
make[1]: Leaving directory '/home/runner/work/go-grpc-middleware/go-grpc-middleware'
cannot detected proto changes or other not formatted files
/bin/bash: run: command not found
M	grpctesting/gogotestpb/fields.pb.go
M	grpctesting/testpb/test.pb.go
Please commit or stash them.
```

This is super weird cannot detected proto changes or other not formatted files - looks like goimports does not format anything at the end :thinking: 

2. Flaky time based tests:  FAIL: TestRetrySuite/TestServerStream_PerCallDeadline_Succeeds (0.54s)

Some work to either mock time or increase timeouts ... Will look tomorrow if you want to help feel free to propose PR to this PR @https://github.com/yashrsharma44 I have time off tomorrow kind of. (:

**Comment by irridia** [comment]
`go mod` behaves and I can import and use the logging. \o/

I feel your pain on the tests...

**Comment by bwplotka** [comment]
Should be good to go - will work on retry flakinesses in separate PR. @yashrsharma44 @johanbrandhorst  (:

**Comment by yashrsharma44** [comment]
Thanks @bwplotka on working on this PR! I will try to make up for the lost time :nerd_face:

**Comment by irridia** [comment]
Can you cut a new tag on branch v2, perhaps `v2.0.0-rc.2`? It'll make mod updates automatic-er.vi g

**Comment by johanbrandhorst** [comment]
I've created a new release: https://github.com/grpc-ecosystem/go-grpc-middleware/releases/tag/v2.0.0-rc.2

**Comment by yashrsharma44** [comment]
Hey @irridia! Can you share how did you correctly install v2 branch?

**Comment by irridia** [comment]
> I've created a new release: https://github.com/grpc-ecosystem/go-grpc-middleware/releases/tag/v2.0.0-rc.2

Thanks. Would you mind also creating a `providers/zerolog/v2.0.0-rc.2` tag so that the zerolog portion of v2 is "released"?

**Comment by irridia** [comment]
> Hey @irridia! Can you share how did you correctly install v2 branch?

Importing at the `github.com/grpc-ecosystem/go-grpc-middleware/v2` path should trigger `go mod tidy` to see the "/v2" suffix and pull the latest tag from the repo of the form `v2.*`. In this case, that will be `v2.0.0-rc.2`.

Generally I just delete the line from go.mod, and run `go mod tidy` to re-perform the dependency match. Though sometimes I need to `go clean -modcache`.

The same method applies when separately importing the log providers at `github.com/grpc-ecosystem/go-grpc-middleware/providers/zerolog/v2`. Right now there are no latest tags for zerolog, so to import it you'd have to find the actual commit hash of the latest v2 merge (`880c1749170035e5a54794546dbfe0d45d10a11d` in this case) and add it manually to go.mod.

As an example, from a fresh `go mod init`, `go mod tidy`, tidy will error out but it creates a go.mod file with this line: ```github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.0.0-rc.2```

Then to get zerolog (for example) importing, manually add this line to `go mod`: ```github.com/grpc-ecosystem/go-grpc-middleware/providers/zerolog/v2 880c1749170035e5a54794546dbfe0d45d10a11d```

Once you run `go mod tidy`, go.mod will be rewritten as: ```	github.com/grpc-ecosystem/go-grpc-middleware/providers/zerolog/v2 v2.0.0-rc.0.0.20200618173137-880c17491700
	github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.0.0-rc.2
```

**Comment by yashrsharma44** [comment]
Thank you so much @irridia ! This was a pretty cool. I googled a lot to find similar issues, but I didn't get any. But the explanation was really awesome, thanks!

**Comment by johanbrandhorst** [comment]
> > I've created a new release: https://github.com/grpc-ecosystem/go-grpc-middleware/releases/tag/v2.0.0-rc.2
> 
> Thanks. Would you mind also creating a `providers/zerolog/v2.0.0-rc.2` tag so that the zerolog portion of v2 is "released"? Also, if there were changes to the other log providers, individual tags for those would be a good idea IMHO.

Done https://github.com/grpc-ecosystem/go-grpc-middleware/releases/tag/providers%2Fzerolog%2Fv2.0.0-rc.2. I have only tagged zerolog.

**Comment by irridia** [comment]
Thanks! `go mod init ; go mod tidy` now does everything automatically with the .../v2 imports. I'll start rolling v2 into our services and see how it goes.
```
	github.com/grpc-ecosystem/go-grpc-middleware/providers/zerolog/v2 v2.0.0-rc.2
	github.com/grpc-ecosystem/go-grpc-middleware/v2 v2.0.0-rc.2
```

**Pull request merged**